include/psa/crypto_values.h

```diff
@@ -1269,6 +1278,74 @@
 #define PSA_ALG_IS_RANDOMIZED_ECDSA(alg)                                \
     (PSA_ALG_IS_ECDSA(alg) && !PSA_ALG_ECDSA_IS_DETERMINISTIC(alg))
 
+/** Edwards-curve digital signature algorithm without prehashing (PureEdDSA),
+ * using standard parameters.
+ *
+ * Contexts are not supported in the current version of this specification
+ * because there is no suitable signature interface that can take the
+ * context as a parameter. A future version of this specification may add
+ * suitable functions and extend this algorithm to support contexts.
+ *
+ * PureEdDSA requires an elliptic curve key on a twisted Edwards curve.
+ * In this specification, the following curves are supported:
+ * - #PSA_ECC_FAMILY_TWISTED_EDWARDS, 255-bit: Ed25519 as specified
+ *   in RFC 8032.
+ *   The curve is Edwards25519.
+ *   The hash function used internally is SHA-512.
+ * - #PSA_ECC_FAMILY_TWISTED_EDWARDS, 448-bit: Ed448 as specified
+ *   in RFC 8032.
+ *   The curve is Edwards448.
+ *   The hash function used internally is the first 114 bytes of the
+ *   SHAKE256 output, with
+ *   `dom4(1, "") = ASCII("SigEd448") || 0x01 0x00`
+ *   prepended to the input.
+ *
+ * This algorithm can be used with psa_sign_message() and
+ * psa_verify_message(). Since there is no prehashing, it cannot be used
+ * with psa_sign_hash() or psa_verify_hash().
+ *
+ * The signature format is the concatenation of R and S as defined by
+ * RFC 8032 §5.1.6 and §5.2.6 (a 64-byte string for Ed25519, a 114-byte
+ * string for Ed448).
+ */
+#define PSA_ALG_PURE_EDDSA                      ((psa_algorithm_t)0x06000800)
+
+#define PSA_ALG_HASH_EDDSA_BASE                 ((psa_algorithm_t)0x06000900)
+#define PSA_ALG_IS_HASH_EDDSA(alg)              \
+    (((alg) & ~PSA_ALG_HASH_MASK) == PSA_ALG_HASH_EDDSA_BASE)
+
+/** Edwards-curve digital signature algorithm with prehashing (HashEdDSA),
+ * using SHAKE256 and the Edwards448 curve.
+ *
+ * See #PSA_ALG_PURE_EDDSA regarding context support and the signature format.
+ *
+ * This algorithm is Ed25519 as specified in RFC 8032.
+ * The curve is Edwards25519.
+ * The input is first hashed with SHA-512.
+ * The hash function used internally is SHA-512, with
+ * `dom2(0, "") = ASCII("SigEd25519 no Ed25519 collisions") || 0x00 0x00`
+ * prepended to the input.
+ */
+#define PSA_ALG_ED25519PH                               \
+    (PSA_ALG_HASH_EDDSA_BASE | (PSA_ALG_SHA_512 & PSA_ALG_HASH_MASK))
+
+/** Edwards-curve digital signature algorithm with prehashing (HashEdDSA),
+ * using SHAKE256 and the Edwards448 curve.
+ *
+ * See #PSA_ALG_PURE_EDDSA regarding context support and the signature format.
+ *
+ * This algorithm is Ed448 as specified in RFC 8032.
+ * The curve is Edwards448.
+ * The input is first hashed by taking the first 64 bytes of the SHAKE256
```

Comment:
Same comment as above - should the caller do the hashing, or does the implementation? - or does this depend on whether sign_message or sign_hash is called?

Reply:
As a user, I would assume that:
* `sign_message` is the all-in-one operation for message `m`
* `sign_hash` on an EdxxxPH algorithm takes in the prehashed message, i.e. `PH(m)` as defined in RFC 8032


But I can see an argument for making `sign_hash` take the fully-hashed input as well (in this case this would be `SHA512(dom2(F, C) || R || A || PH(M))`). Defining the API that way would probably make a driver implementation underneath easier, but is not as intuitive for a user.

Reply:
As with other hash-and-sign algorithms, you can either hash the message first and call {sign,verify}_hash, or let the all-in-one signature take care of that with {sign,verify}_message. What PSA calls the hashing step is what EdDSA calls the prehash. The hashing step is independent of the key. The hash calculation used internally as part of the EdDSA calculation is irrelevant to the user, so it isn't exposed through the API.

Reply:
Is it our place (at least in these docs) to discuss why this is done (at least the possible security implications) of the difference between directly signing and the hash and sign? At least the authors argue of protecting against possible weaknesses in the hash function with their construct, and those protections only work by signing the entire message.

Somewhat related, would we want to have an API that would allow the entire message to be signed, but allow the internal hashing to be done in pieces? Or would users just be expected to hash the message first themselves in this case? Although, if someone wished for the possible security benefit of signing the entire message, they would need to have internal access to how the hash that is happening under the hood happens.

Reply:
I'm not a fan of limiting HashEdDSA to `psa_(sign|verify)_hash()`, on the justification that an application which can call `psa_(sign|verify)_message()` (i.e. it has the entire message in a single buffer) should use PureEdDSA because it has better security properties. Interoperability may require the use of HashEdDSA, and we should not make it harder for an application than necessary by denying the use of `psa_(sign|verify)_message()`.

If/when we provide a multi-part asymmetric signature operation API, there is a strong case to make it not support algorithms which require two passes of the input. Although that would make for an interesting debate about whether to support verification of PureEdDSA (only needs one hash computation on the message data), but not signature (which requires two sequentially dependent passes)?